### DISCUSSION TITLE
Can multiple identical response be simpler?

### DISCUSSION THREAD

**Discussion by Robin528919** [fastapi/fastapi#11827] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from aioredis import Redis
from fastapi import APIRouter
from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession
from model.user import Users as UsersModel
from db.redis import get_redis
from db.session import get_session
from schemas.base import BaseRespT, BaseRespSuccess, BaseRespFail
from schemas.sub_user import *
from curd.curd_sub_user import curd_sub_user
from common.deps import get_current_user

router = APIRouter(
    prefix="/sub-user",
    tags=["子用户"]
)


# 添加子用户
@router.post("", response_model=Union[BaseRespSuccess, BaseRespFail], summary="添加子用户")
async def add(
        sub_user_info: SubUserCreate,
        user: UsersModel = Depends(get_current_user),
        db: AsyncSession = Depends(get_session),
):
    if user_info := await curd_sub_user.create(db, obj_in=sub_user_info.model_dump(), creator_id=user.id):
        return BaseRespT(data=user_info)
    return BaseRespT(msg="添加失败")


# 删除子用户
@router.delete("", response_model=Union[BaseRespSuccess, BaseRespFail], summary="删除子用户")
async def delete(
        sub_user_info: SubUserDelete,
        user: UsersModel = Depends(get_current_user),
        db: AsyncSession = Depends(get_session),
):
    if await curd_sub_user.delete(db, _id=sub_user_info.id, deleter_id=user.id):
        return BaseRespT(msg="删除成功")
    return BaseRespT(msg="删除失败")


# 禁用子用户
@router.put("", response_model=Union[BaseRespSuccess, BaseRespFail], summary="禁用子用户")
async def disable():
    pass


# 释放端口
@router.put("", response_model=Union[BaseRespSuccess, BaseRespFail], summary="释放端口")
async def release():
    pass
```

### Description

Can multiple identical response be simpler?
Can I merge multiple response_model=Union[BaseRespSuccess, BaseRespFail] into a blueprint?

### Operating System

macOS

### Operating System Details

(Aboluo) zhangpengqiang@zhangpeangdeMBP Aboluo % pip show fastapi
Name: fastapi
Version: 0.103.0
Summary: FastAPI framework, high performance, easy to learn, fast to code, ready for production
Home-page: 
Author: 
Author-email: Sebastián Ramírez <tiangolo@gmail.com>
License: 
Location: /Users/zhangpengqiang/opt/anaconda3/envs/Aboluo/lib/python3.8/site-packages
Requires: pydantic, starlette, typing-extensions
Required-by: 

### FastAPI Version

0.103.0

### Pydantic Version

 2.5.3

### Python Version

Python 3.8.19

**Comment by YuriiMotov** [comment]
`Union` is not supported for `response_model` parameter: https://fastapi.tiangolo.com/tutorial/response-model/#invalid-return-type-annotations

You can overcome it by using [pydantic discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions). Here is an example:
```
from typing import Annotated, Literal, TypeAlias, Union
from fastapi import FastAPI
from pydantic import BaseModel, Field


app = FastAPI()


class ErrorResponse(BaseModel):
    response_type: Literal["ErrorResponse"] = "ErrorResponse"
    msg: str


class SuccessfulResponse(BaseModel):
    response_type: Literal["SuccessfulResponse"] = "SuccessfulResponse"
    data: str


AnyResponse: TypeAlias = Annotated[
    Union[ErrorResponse, SuccessfulResponse], Field(discriminator="response_type")
]


@app.get("/", response_model=AnyResponse)
def get_index(v: int):
    if v == 1:
        return SuccessfulResponse(data="some data")
    return ErrorResponse(msg="error description")

```

**Comment by Robin528919** [comment]
What I hope is

router = APIRouter(
    prefix="/sub-user",
    response_model=Union[BaseRespSuccess, BaseRespFail],
    tags=["子用户"]
)

No need to specify each

**Reply by YuriiMotov** [reply]
According to the link I mentioned above, `response_model` should be a **valid Pydantic type**. `Union[BaseRespSuccess, BaseRespFail]` is not.

To make it valid Pydantic type you need to specify **discriminator**.
If you don't want to change models like in the example above, you can also use **callable descriminator**:

```
from typing import Annotated, TypeAlias, Union
from fastapi import FastAPI
from pydantic import BaseModel, Discriminator, Field, Tag

app = FastAPI()

class ErrorResponse(BaseModel):
    msg: str

class SuccessfulResponse(BaseModel):
    data: str

AnyResponse: TypeAlias = Annotated[
    Union[
        Annotated[ErrorResponse, Tag("error")], 
        Annotated[SuccessfulResponse, Tag("success")],
    ],
    Field(
        discriminator=Discriminator(lambda x: "error" if hasattr(x, "msg") else "success")
    )
]

@app.get("/", response_model=AnyResponse)
def get_index(v: int):
    if v == 1:
        return SuccessfulResponse(data="some data")
    return ErrorResponse(msg="error description")
```
